### PR TITLE
Potential fix for HeaderMap issues

### PR DESCRIPTION
**Pull request by jaraco**
**What kind of change does this PR introduce?**
  - [X] bug fix
  - [ ] feature
  - [ ] docs update
  - [ ] tests/coverage improvement
  - [ ] refactoring
  - [ ] other

**What is the related issue number (starting with `#`)**

#1827

# Discussion

I'm not sure this fix is the right one. I did not attempt to resolve what values _should_ be allowed for keys to a `HeaderMap`, only to address the issue that _if_ they're bytes, they're treated duck-like strings. Perhaps the right thing to do is just crash on non-text keys and require the callers to decode any bytes values.

I've called out `None` values as a special case, as tests fail without them. Perhaps that should be an error too, but for now, because the tests demand it, they're supported.

### CODE REVIEW

File: cherrypy/lib/httputil.py [hunk 1]
```diff
@@ -450,12 +450,15 @@ class CaseInsensitiveDict(KeyTransformingDict):
...
-    Each key is changed on entry to str(key).title().
+    Each key is changed on entry to title case.
     """
 
     @staticmethod
     def transform_key(key):
-        return str(key).title()
+        if key is None:
+            # TODO: why?
+            return 'None'
```
Comment (webknjaz) [inline cherrypy/lib/httputil.py:401]: This should probably be
```suggestion
            raise ValueError
```

**Comment by webknjaz** [comment]
@jaraco I think we should make the public API of HeaderMap compatible with [multidict] which would allow us to use it as a speed-up option.

[multidict]: https://pypi.org/project/multidict/

File: cherrypy/lib/httputil.py (hunk 1 above)
Comment (jaraco) [reply cherrypy/lib/httputil.py:401]: Have you tried that? Lots of tests fail when key is `None`. I'd like to see that issue fixed too, but it's outside the scope of the reported issue.

**Comment by jaraco** [comment]
> I think we should make the public API of HeaderMap compatible with [multidict](https://pypi.org/project/multidict/) which would allow us to use it as a speed-up option.

I appreciate the reference. That seems reasonable, but I'm pretty sure that class has a public interface, so you'll need to support both interfaces for some time. And it's not obvious to me it will be possible. In any case, that sounds like a feature request separate from the issue at hand. Feel free to file one (and implement it).

File: cherrypy/lib/httputil.py (hunk 1 above)
Comment (webknjaz) [reply cherrypy/lib/httputil.py:401]: Oh. Then this is worth mentioning in a comment.
This should probably be
```suggestion
            # NOTE: Setting `raise ValueError` here causes many tests
            # to fail so before refactoring we'll need to fix that.
            return 'None'
```

Comment (jaraco) [reply cherrypy/lib/httputil.py:401]: Right. I was trying to be more concise with `TODO: why?`... and let the failing tests tell the rest of the story until a ticket could be filed about it. With this change, we have both a TODO and a NOTE. That seems excessive, but thanks for the suggestion.

Comment (jaraco) [reply cherrypy/lib/httputil.py:401]: In b07830a1, I've updated the TODO to reference a new issue (#1830).

**Pull request merged**